### ISSUE TITLE
apt update_cache gives error on ubuntu for Ansible 2.0.0.2 

### ISSUE THREAD

**Issue by xuchen81** [ansible/ansible-modules-core#2951]
##### ISSUE TYPE

Bug Report
##### COMPONENT NAME

apt module
##### ANSIBLE VERSION

N/A
##### SUMMARY

The simple following command:

```
- name: Update apt cache.
  apt: update_cache=true
  become: true
```

Throws an error on AWS Ubuntu instances.

fatal: [xx.xx.xxx.x]: FAILED! => {"changed": false, "failed": true, "module_stderr": "", "module_stdout": "Traceback (most recent call last):\r\n  File \"/home/ubuntu/.ansible/tmp/ansible-tmp-1454536847.17-86172912102079/apt\", line 2630, in <module>\r\n    main()\r\n  File \"/home/ubuntu/.ansible/tmp/ansible-tmp-1454536847.17-86172912102079/apt\", line 603, in main\r\n    cache = apt.Cache()\r\n  File \"/usr/lib/python2.7/dist-packages/apt/cache.py\", line 107, in **init**\r\n    self.open(progress)\r\n  File \"/usr/lib/python2.7/dist-packages/apt/cache.py\", line 151, in open\r\n    self._cache = apt_pkg.Cache(progress)\r\nSystemError: E:Encountered a section with no Package: header, E:Problem with MergeList /var/lib/apt/lists/us-west-2.ec2.archive.ubuntu.com_ubuntu_dists_trusty-updates_universe_i18n_Translation-en, E:The package lists or status file could not be parsed or opened.\r\n", "msg": "MODULE FAILURE", "parsed": false}

However, when I ssh into the instance, and do sudo apt-get update manually, there is no problem at all. And then, if I ran the task again, the problem will go away, I guess because I did the manual update on the instance.

Any ideas?

Thanks

**Comment by noqcks** [comment]
I have ran into this problem before too. Someone should look into it. @mgwilliams

**Comment by xuchen81** [comment]
Any updates on this?

**Comment by aedades** [comment]
+1

**Comment by ghost** [comment]
+1

If i'm not mistaken, by the traceback it looks like this error is occurring within the Python apt module that Ansible uses?

**Comment by alexahn** [comment]
+1
Running `sudo rm -vf /var/lib/apt/lists/*` gets around the problem.

**Comment by adamrmelnyk** [comment]
+1
I've stopped using the apt module entirely because of this. Running:

``` yaml
shell: apt-get update
```

is **much** more reliable, so it seems the issue stems from the python apt module like others have pointed out. That being said you will get annoying warnings like this: 

``` sh
[WARNING]: Consider using apt-get module rather than running apt-get
```

**Comment by llbbl** [comment]
```
sudo apt-get clean 
sudo apt-get update
```

this fixed it for me, running it on target machine

**Comment by xuchen81** [comment]
@jimi-c  any updates on this?

**Comment by dohoangkhiem** [comment]
+1
Ansible 1.9.4
Exactly the same error, doing 'apt-get update' command succeeded while always failed with 3 retries of apt module

Removing /var/lib/apt/lists/\* could resolve this but as command 'sudo apt-get update' works without that cleanup so seems this is due to Ansible apt module.

Right now I have to add a retry with command in case apt module failed

```
- name: apt update
  apt: update_cache=yes cache_valid_time=3600
  register: apt_result
  until: apt_result|success
  retries: 3
  delay: 1
  sudo: yes
  ignore_errors: yes

- name: retry if needed using command apt-get update
  command: apt-get update
  sudo: yes
  when: apt_result|failed
```

**Comment by xuchen81** [comment]
@jimi-c 
Problem remains for **Ansible Version 2.0.1.0**, any future plans on this bug?

**Comment by stemail23** [comment]
+1

**Comment by tmpvar** [comment]
I've also noticed that if you pass in a `--limit=aSingleHost` it generally works.

using @dohoangkhiem's retry and then `command` trick, I went from ~90% failure to < 1% failure over 130 machines.

**Comment by fwisehc** [comment]
+1 yes have this problem also. needs to be fixed

```
- name: hack to resolve Problem with MergeList Issue
  sudo: true
  shell: 'find /var/lib/apt/lists -maxdepth 1 -type f -exec rm -v {} \;'

- apt: update_cache=yes
```

**Comment by xuchen81** [comment]
So it still happens in Ansible 2.0.2.0.  Does anisble not even care about this bug? @jimi-c

**Comment by zornhsu** [comment]
+1 
$ ansible --version
ansible 2.0.0.2

**Comment by anmolnagpal** [comment]
Hi,

I am integrating Slack module with Ansible getting same type fo error 

```
TASK [redis : slack] ***********************************************************
fatal: [XX.XX.XX.XX -> localhost]: FAILED! => {"changed": false, "failed": true, "module_stderr": "sudo: a password is required\n", "module_stdout": "", "msg": "MODULE FAILURE", "parsed": false}
 [ERROR]: Could not create retry file './ssh/retry/redis-all.retry'. The error was: [Errno 2] No such file or directory: './ssh/retry/redis-all.retry'
```

**Comment by xxthegonzxx** [comment]
+1

**Comment by ms-ati** [comment]
+1

**Comment by rubensmets** [comment]
+1 btw: I'm having this issue running from osx

**Comment by igr** [comment]
The same with ansible 2.1; I use @dohoangkhiem thing, but i have few places in my playbook where I deal with the apt (update cache, install java, install docker). So I am thinking to use `shell` instead.

**Comment by arjun810** [comment]
Also using dohoangkhiem's workaround, but it'd be great if this worked reliably.

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by xuchen81** [comment]
Updates??

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by siwyd** [comment]
~~+1 on 1.9.6~~

Correction: I had the same error, but it was related to trying to run the same apt task on the same host multiple times (at the same time) if I recall correctly.

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by akamensky** [comment]
- Ansible 2.1 from pip

I guess since Ansible has been acquired by RH, we can just forget about apt module support, or have to fix it ourselves.

**Comment by llbbl** [comment]
@akamensky well if they are going to drop support for ubuntu looks like we will have to drop support for ansible.

**Comment by xenithorb** [comment]
Since it's so important to you why not fix it and submit a PR?

Clearly this is some edge case with the images provided by ubuntu on amazon (no surprise there, their docker containers suck, too) 

Why not just mitigate this with a task to clear out `/var/lib/apt/lists/*` if you can detect amazon AND ubuntu from facts?

**Comment by akamensky** [comment]
@xenithorb Edge case for any Ubuntu/Debian installation in my case. I've tried Debian 7 netinstall, Ubuntu 14.04 ISO install, I actually haven't tried Amazon ones. But from what you say it look like the issue is also there.

As for making PR -- I did, those are never even looked at.

**Comment by xenithorb** [comment]
[Your PR](https://github.com/ansible/ansible/pull/17115/files) has only been published for a month and it has _nothing_ to do with this issue. Which I might add isn't even a bug fix, but an enhancement. (It even links back to a feature_idea tagged issue ansible/ansible#11175). Are you really that surprised they're probably prioritizing bugs over feature ideas?

For a serious bug like this I'm sure if you fixed it you could poke the right people (politely) and perhaps even get on IRC and get something done.

I honestly haven't seen this bug myself, though I admittedly don't touch the debian family side of the house often. When I do though I've noticed stability issues with apt (even sometimes no responsibility of the module itself, the back-end fails contacting a mirror and the logic to recover from that is just bad on their (apt's) behalf. I usually mitigate that with an `until:` + `retries:` loop and let it do its thing.

**Comment by adichiru** [comment]
+1 On Ubuntu 16.04.1 LTS with ansible 2.0.0.2 and python 2.7

Has nothing to do with Amazon.

sudo rm -rf /var/lib/apt/lists/\* solve the issues momentarily for me too.

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by adichiru** [comment]
Do we know who is the maintainer for this? The issue was opened 8 months ago and constantly reported since and it is a core module as we all know. I'd like to know if the maintainer is still available for fixes or we need to find another way.

**Comment by lukeorland** [comment]
+1
- Ubuntu 16.10 remote host
- macOS ansible master
  
  ```
  $ ansible-playbook --version
  ansible-playbook 2.1.1.0
  ```
  
  ```
  $ python --version
  Python 2.7.10
  ```

I attempted the workaround suggested [above](https://github.com/ansible/ansible-modules-core/issues/2951#issuecomment-218327951), but the directory doesn't exist

```
FAILED! => {"changed": true, "cmd": "find /var/lib/apt/lists -maxdepth 1 -type f -exec rm -v {} \\;", "delta": "0:00:00.011281", "end": "2016-10-20 22:34:26.408833", "failed": true, "rc": 1, "start": "2016-10-20 22:34:26.397552", "stderr": "find: /var/lib/apt/lists: No such file or directory", "stdout": "", "stdout_lines": [], "warnings": []}
```

The @dohoangkhiem fix was not successful:

```
TASK [apt] *********************************************************************
FAILED - RETRYING: TASK: apt (3 retries left).
FAILED - RETRYING: TASK: apt (2 retries left).
FAILED - RETRYING: TASK: apt (1 retries left).
fatal: [yogax]: FAILED! => {"changed": false, "cmd": "apt-get update", "failed": true, "msg": "[Errno 2] No such file or directory", "rc": 2}
...ignoring

TASK [retry if needed using command apt-get update] ****************************
fatal: [yogax]: FAILED! => {"changed": false, "cmd": "apt-get update", "failed": true, "msg": "[Errno 2] No such file or directory", "rc": 2}
```

**Comment by igr** [comment]
This is one of the most annoying issues you can face with ansible, especially since you bump to it very early: practically your very first script will install something using apt. I wasted too much time just to make crazy workarounds that worked only for specific target OS.

**Comment by adichiru** [comment]
I agree @igorspasic. I was assuming Ansible is more mature...

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by davey-dev** [comment]
This suddenly just started happening to me on Ubuntu 14.04. Was working flawlessly before now.

**Comment by jctanner** [comment]
Related:
http://askubuntu.com/questions/561276/my-repos-became-corrupted-unable-to-parse-package-file
https://answers.launchpad.net/ubuntu/+source/apt/+question/267919
https://bugs.launchpad.net/ubuntu/+source/apt/+bug/346386

**Comment by jctanner** [comment]
I am able to reproduce the bug by injecting noise into one of the package list files ...

```
- hosts: u1610
  user: vagrant
  gather_facts: False
  vars:
    pfile: /var/lib/apt/lists/archive.ubuntu.com_ubuntu_dists_trusty-updates_main_binary-amd64_Packages
  tasks:
    - name: screw up a package file
      shell: head -n 100000 /dev/urandom >> {{ pfile }}
      become: true

    - name: Update apt cache.
      apt: update_cache=true
      become: true
```

```
fatal: [u1610]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "invocation": {
        "module_name": "apt"
    }, 
    "module_stderr": "OpenSSH_7.2p2, OpenSSL 1.0.2h-fips  3 May 2016\r\ndebug1: Reading configuration data /home/jtanner/.ssh/config\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: /etc/ssh/ssh_config line 58: Applying options for *\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 15640\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 2\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 0\r\nShared connection to u1610 closed.\r\n", 
    "module_stdout": "\r\nTraceback (most recent call last):\r\n  File \"/tmp/ansible_qg6gD1/ansible_module_apt.py\", line 935, in <module>\r\n    main()\r\n  File \"/tmp/ansible_qg6gD1/ansible_module_apt.py\", line 825, in main\r\n    cache = apt.Cache()\r\n  File \"/usr/lib/python2.7/dist-packages/apt/cache.py\", line 107, in __init__\r\n    self.open(progress)\r\n  File \"/usr/lib/python2.7/dist-packages/apt/cache.py\", line 151, in open\r\n    self._cache = apt_pkg.Cache(progress)\r\nSystemError: E:Unable to parse package file /var/lib/apt/lists/archive.ubuntu.com_ubuntu_dists_trusty-updates_main_binary-amd64_Packages (1), E:The package lists or status file could not be parsed or opened.\r\n", 
    "msg": "MODULE FAILURE"
}

```

**Comment by jctanner** [comment]
My final testcase used to write #5448 ...

```
- hosts: u1610
  user: vagrant
  gather_facts: False
  vars:
    pfile: /var/lib/apt/lists/archive.ubuntu.com_ubuntu_dists_trusty-updates_main_binary-amd64_Packages
    #pfile: /var/lib/apt/lists/*Packages
  tasks:

    - name: fix the cache
      shell: apt-get update
      ignore_errors: true
      with_sequence: count=2
      become: true

    - name: screw up a package file
      shell: head -n 100000 /dev/urandom >> {{ pfile }}
      become: true

    #- name: fix the cache
    #  shell: apt-get update
    #  ignore_errors: True
    #  with_sequence: count=2
    #  become: true

    - name: check if the cache is broken
      shell: apt-cache search vim
      ignore_errors: true
      become: true
      register: cache_check

    - name: make sure the cache is broken
      assert:
        that:
            - 'cache_check.rc != 0'

    - name: Update apt cache.
      apt: update_cache=true
      become: true
```

**Comment by Sumant-s-p** [comment]
I am getting this error when I run apt-get update through ansible

FAILED! => {"changed": false, "failed": true, "module_stderr": "Shared connection to  closed.\r\n", "module_stdout": "sudo: unable to resolve host ip-\r\nTraceback (most recent call last):\r\n  File \"/tmp/ansible_gvpRK4/ansible_module_apt.py\", line 909, in <module>\r\n    main()\r\n  File \"/tmp/ansible_gvpRK4/ansible_module_apt.py\", line 799, in main\r\n    cache = apt.Cache()\r\n  File \"/usr/lib/python2.7/dist-packages/apt/cache.py\", line 107, in __init__\r\n    self.open(progress)\r\n  File \"/usr/lib/python2.7/dist-packages/apt/cache.py\", line 153, in open\r\n    self._records = apt_pkg.PackageRecords(self._cache)\r\nSystemError: E:Could not open file /var/lib/apt/lists/archive.ubuntu.com_ubuntu_dists_trusty_universe_binary-amd64_Packages - open (2: No such file or directory), E:Could not open file /var/lib/apt/lists/archive.ubuntu.com_ubuntu_dists_trusty_restricted_binary-amd64_Packages - open (2: No such file or directory), E:Could not open file /var/lib/apt/lists/archive.ubuntu.com_ubuntu_dists_trusty_main_binary-amd64_Packages - open (2: No such file or directory)\r\n", "msg": "MODULE FAILURE"}
	to retry, use: --limit @/var/lib/jenkins/workspace/create-base-ami-ehqv2/base-ami-creation/ehqv2/site.retry

**Comment by rommsen** [comment]
@jctanner please reopen this. it is not fixed
with ansible 2.2.1.0 we get

```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: SystemError: E:Datei /var/lib/apt/lists/ftp.hosteurope.de_mirror_archive.ubuntu.com_dists_trusty_universe_binary-i386_Packages konnte nicht geöffnet werden. - open (2: No such file or directory)
failed: [ipm_promotion_stage] (item=[u'postgresql-9.3', u'postgresql-client-9.3']) => {"failed": true, "item": ["postgresql-9.3", "postgresql-client-9.3"], "module_stderr": "Traceback (most recent call last):\n File \"/tmp/ansible_6GV6h7/ansible_module_apt.py\", line 903, in <module>\n main()\n File \"/tmp/ansible_6GV6h7/ansible_module_apt.py\", line
799, in main\n    cache = apt.Cache()\n  File \"/usr/lib/python2.7/dist-packages/apt/cache.py\", line 107, in __init__\n    self.open(progress)\n  File \"/usr/lib/python2.7/dist-packages/apt/cache.py\", line 153, in open\n    self._records = apt_pkg.PackageRecords(self._cache)\nSystemError: E:Datei /var/lib/apt/lists/ftp.hosteurope.de_mirror_archive.ubuntu.com_dists_trusty_universe_binary-i386_Packages konnte nicht geöffnet werden. - open (2: No such file or directory)\n", "module_stdout": "", "msg": "MODULE FAILURE"}
```
There is no problem when doing this with apt-get directly. The files are there.
We do `apt-get clean` and `rm -rf /var/lib/apt/lists/*` when the playbook starts manually but no luck.

This is not breaking on every playbook run but completely random during different tasks that use the apt module.

**Comment by casept** [comment]
This bug is also affecting me. What's worse, it's not even an issue affecting a particular target OS. I tried executing this on `ubuntu xenial`, `ubuntu trusty`, `debian jessie` and `debian testing`.

```
fatal: [debian_testing]: FAILED! => {"changed": false, "failed": true, "msg": "Failed to update apt cache."}
fatal: [debian_jessie]: FAILED! => {"changed": false, "failed": true, "msg": "Failed to update apt cache."}
fatal: [ubuntu_xenial]: FAILED! => {"changed": false, "failed": true, "msg": "Failed to update apt cache."}
fatal: [ubuntu_trusty]: FAILED! => {"changed": false, "failed": true, "msg": "Failed to update apt cache."}
        to retry, use: --limit @/home/david/dev/ansible-playbooks/docker.retry
 ____________
< PLAY RECAP >
 ------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

debian_jessie              : ok=1    changed=0    unreachable=0    failed=1   
debian_testing             : ok=1    changed=0    unreachable=0    failed=1   
ubuntu_trusty              : ok=1    changed=0    unreachable=0    failed=1   
ubuntu_xenial              : ok=1    changed=0    unreachable=0    failed=1  
```
Ansible version:
```
david@T400:~/dev/ansible-playbooks$ ansible-playbook --version
ansible-playbook 2.2.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

**Comment by cultcom** [comment]
same with me:

ansible 2.2.1.0 (stable-2.2 4e257fad84)

**Comment by fcanela** [comment]
Facing the same problem when updating an Ubuntu 14.04 vagrant box. Ansible version 2.2.1.0 too.

**Comment by xenithorb** [comment]
@cultcom @fcanela I don't believe this repo is active anymore as of 2.3. If you have a reproducible test case, may I suggest making a bug over at [ansible/ansible](https://github.com/ansible/ansible)?

**Comment by yuriybabenko** [comment]
Also struggling with this issue.

- OSX 10.12.3
- Ansible 2.2.1.0

Trying to run this recipe: https://github.com/cthos/ansible-homestead-selenium/blob/master/selenium_homestead.yml 

Fails here:

`TASK [Install chromium-browser] ************************************************
fatal: [homestead]: FAILED! => {"changed": false, "failed": true, "msg": "Failed to update apt cache."}`

// edit - there's a suggested workaround in https://github.com/ansible/ansible/issues/18987#issuecomment-268887567

**Comment by zhangchiqing** [comment]
+1

Still have this issue with Ubuntu 14.04 and ansible-playbook 2.2.1.0, but here is a reliable workaround. Thanks @adamrmelnyk and @llbbl 

```
- name: apt-get clean
  shell: apt-get clean
  become: true
  become_user: root

- name: apt-get update
  shell: apt-get update
  become: true
  become_user: root
```

Update: It only works when running against hosts one by one, meaning you have to run ansible-playbook with `--limit`. 
```
ansible-playbook -i hosts/vagrant provision.yml --limit 10.100.181.202
```

**Comment by jovinbm** [comment]
Ubuntu 14.04

I spent hours trying to find a solution for this. The one thing that fixed it for me was to [remove invalid ppas](http://askubuntu.com/a/310) from my lists. Be sure to check your `/etc/apt/sources.list.d` directory. I had noticed that `shell: apt-get update` was showing some 404s on some of them. Maybe ansible apt module considers these as errors? Doing `apt-get update` while inside vm will just show these as warnings and proceed with the update.

 I guess I got them added as a result of trial and error on various open source ansible modules. 
(Some modules allow you to use a version as a variable name and they will dynamically append that without validating it which ends up being unreachable)

**Comment by thedumbtechguy** [comment]
> Ubuntu 16.10
> 
> ansible 2.3.0.0
> python version = 2.7.12+ (default, Sep 17 2016, 12:08:02) [GCC 6.2.0 20160914]
> 

In my case, it run fine the first time, failed consistently after that.

**Comment by thedumbtechguy** [comment]
Thanks to @jovinbm  for pointing me in the right direction.

I was rewriting an aging role for my own use. This role had a dated repository which is what I was trying to replace. 

When I run `apt-get update`, the output showed an error:
```shell
Reading package lists... Done
W: GPG error: http://mirror.netcologne.de/mariadb/repo/10.1/ubuntu yakkety InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY F1656F24C74CD1D8
W: The repository 'http://mirror.netcologne.de/mariadb/repo/10.1/ubuntu yakkety InRelease' is not signed.
N: Data from such a repository can't be authenticated and is therefore potentially dangerous to use.
```

removing the offending lists solved my issue

**Comment by merito** [comment]
To get rid of this error you can add an option to your sources.list (of course if you trust your repository)

> • trusted=yes can be set to indicate that packages from this source are always authenticated even if the Release file is not signed or the signature can't be checked. This disables parts of apt-secure(8) and should therefore only be used in a local and trusted context. trusted=no is the opposite which handles even correctly authenticated sources as not authenticated.

it is going to look like this
```
deb [trusted=yes] http://deb.debian.org/debian jessie main
```
It is not a perfect solution, but good enough for me by the time they fix it.

Source: [sources.list(5) man page](https://manpages.debian.org/jessie/apt/sources.list.5.en.html#THE_DEB_AND_DEB-SRC_TYPES)

**Comment by ralffunk** [comment]
Even though this is closed, I would like to add my experience in case it helps others.

In my case ansibles "apt: update_cache: yes" was failing because the command "sudo apt update" was raising some warnings about a cdrom source that it could not find or wasn't in the expected format. Apparently the ansible "apt" module caught these warnings as errors and proceded to fail the task.

My solution was to remove (or comment) the lines with "cdrom" in "/etc/apt/sources.list". I did this through ansible, using the "lineinfile" module.

My setup:
- local machine: ubuntu 17.10
- target host: ubuntu server LTS 16.04.2
- Ansible 2.4
- Python 3 on both machines

Hope this helps someone.

Added the same comment to https://github.com/ansible/ansible/issues/18987
